src/Microsoft.ML.TimeSeries/RootCauseAnalyzer.cs

```diff
@@ -0,0 +1,764 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.ML.Internal.Utilities;
+
+namespace Microsoft.ML.TimeSeries
+{
+    public class RootCauseAnalyzer
+    {
+        private static double _anomalyRatioThreshold = 0.5;
+        private static double _anomalyDeltaThreshold = 0.95;
+        private static double _anomalyPreDeltaThreshold = 2;
+
+        private RootCauseLocalizationInput _src;
+        private double _beta;
+
+        public RootCauseAnalyzer(RootCauseLocalizationInput src, double beta)
+        {
+            _src = src;
+            _beta = beta;
+        }
+
+        public RootCause Analyze()
+        {
+            return AnalyzeOneLayer(_src);
+        }
+
+        //This is a function for analyze one layer for root cause, we select one dimension with values who contributes the most to the anomaly. For full result, call this function recursively
+        private RootCause AnalyzeOneLayer(RootCauseLocalizationInput src)
+        {
+            RootCause dst = new RootCause();
+            dst.Items = new List<RootCauseItem>();
+
+            DimensionInfo dimensionInfo = SeperateDimension(src.AnomalyDimension, src.AggSymbol);
+            Tuple<PointTree, PointTree, Dictionary<string, Point>> pointInfo = GetPointsInfo(src, dimensionInfo);
+            PointTree pointTree = pointInfo.Item1;
+            PointTree anomalyTree = pointInfo.Item2;
+            Dictionary<string, Point> dimPointMapping = pointInfo.Item3;
+
+            //which means there is no anomaly point with the anomaly dimension or no point under anomaly dimension
+            if (anomalyTree.ParentNode == null)
+            {
+                return dst;
+            }
+            if (dimPointMapping.Count == 0)
+            {
+                return dst;
+            }
+
+            dst.Items.AddRange(LocalizeRootCauseByDimension(anomalyTree, pointTree, src.AnomalyDimension, dimensionInfo.AggDims));
+            GetRootCauseDirectionAndScore(dimPointMapping, src.AnomalyDimension, dst, _beta, pointTree, src.AggType, src.AggSymbol);
+
+            return dst;
+        }
+
+        protected List<Point> GetTotalPointsForAnomalyTimestamp(RootCauseLocalizationInput src)
+        {
+            List<Point> points = new List<Point>();
+            foreach (MetricSlice slice in src.Slices)
+            {
+                if (slice.TimeStamp.Equals(src.AnomalyTimestamp))
+                {
+                    points = slice.Points;
+                }
```

Comment:
Are you guaranteed to find a matching element? Is it going to be unique? 

Is the intention to return the first or the last element?  If it is the first, you can add a break after this statement. If it is the last, you can may be search in reverse and add a break.

In either case, you can perhaps rewrite this code using LINQ. E.g. `return src.Slices.Single(slice => slice.TimeStamp.Equals(src.AnomalyTimestamp))`

Reply:
> Are you guaranteed to find a matching element? Is it going to be unique?
> 
> Is the intention to return the first or the last element? If it is the first, you can add a break after this statement. If it is the last, you can may be search in reverse and add a break.
> 
> In either case, you can perhaps rewrite this code using LINQ. E.g. `return src.Slices.Single(slice => slice.TimeStamp.Equals(src.AnomalyTimestamp))`

the slice might be in the first or in the last, or in the middle. I think your suggestion is great